src/utils/service_builder.rs

```diff
@@ -220,6 +224,16 @@ fn build_providers(
             .into());
         }
 
+        // Check for duplicate provider names.
```

Comment:
This seems to check for duplicate accross all providers as opposed to only providers of the same UUID. Are we allowing that providers of different type have the same name?
I think it's better that way (similarly for key names), that provider names are unique. That would also simplify the schema as `provider_name` alone can now be used to choose a provider.

Reply:
Yes, that's what I was going for. It's a design choice that can be up for debate though.
Personally I'm indifferent about how we scope `provider_name`'s uniqueness.